src/main/java/io/openliberty/tools/intellij/lsp4mp/lsp4ij/operations/diagnostics/LSPDiagnosticAnnotator.java

```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.tools.intellij.lsp4mp.lsp4ij.operations.diagnostics;
+
+import com.intellij.codeInsight.intention.IntentionAction;
+import com.intellij.lang.annotation.AnnotationBuilder;
+import com.intellij.lang.annotation.AnnotationHolder;
+import com.intellij.lang.annotation.ExternalAnnotator;
+import com.intellij.lang.annotation.HighlightSeverity;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiFile;
+import io.openliberty.tools.intellij.lsp4mp.lsp4ij.LSPIJUtils;
+import io.openliberty.tools.intellij.lsp4mp.lsp4ij.LSPVirtualFileWrapper;
+import io.openliberty.tools.intellij.lsp4mp.lsp4ij.LanguageServiceAccessor;
+
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Intellij {@link ExternalAnnotator} implementation which get the current LSP diagnostics for a given file and translate
+ * them into Intellij {@link com.intellij.lang.annotation.Annotation}.
+ */
+public class LSPDiagnosticAnnotator extends ExternalAnnotator<LSPVirtualFileWrapper, LSPVirtualFileWrapper> {
+
+    @Nullable
+    @Override
+    public LSPVirtualFileWrapper collectInformation(@NotNull PsiFile file, @NotNull Editor editor, boolean hasErrors) {
+        // make sure LS is initialized for this document TODO: probably should find a way to do this on file open?
+        LanguageServiceAccessor.getInstance(file.getProject()).getLanguageServers(editor.getDocument(), (capabilities) -> { return true; });
```

Comment:
Maybe the capabilities lambda should check if the LS supports diagnostics (`getDiagnosticProvider()`)

Reply:
I wonder why you wish to do that? Indeed LSP diagnostics are not requested but are notified. It means that it is not the LSP client which request the LSP diagnostics (in this case I agree wiyth you,you need to check this diagnostic capability),but as it is the LSP language server which publish diagnostics, it will fill the diagnostics cache in LSPVirtualFileWrapper.

Reply:
I added this call to make sure the LS are ready by the time we process and apply diagnostics. As mentioned [here](https://github.com/OpenLiberty/liberty-tools-intellij/issues/367#issuecomment-1527667451), the rest of our language servers don't show diagnostics on opening the file for the first time (except for MP), and have to trigger the LS to load (like with a hover) documented in #329.

Also it seems checking diagnostic capabilities does not help in this case.